File: sigil-go/pkg/models/recipe.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type Recipe struct {
	ID          uuid.UUID    `json:"id"          db:"id"`
	Name        string       `json:"name"        db:"name"`
	Summary     *string      `json:"summary"     db:"summary"`
	Servings    *int         `json:"servings"    db:"servings"`
	PrepTime    *string      `json:"prepTime"    db:"prep_time"`
	SourceURL   *string      `json:"sourceUrl"   db:"source_url"`
	Ingredients []Ingredient `json:"ingredients" db:"ingredients"`
	Steps       []string     `json:"steps"       db:"steps"`
	CreatedAt   time.Time    `json:"createdAt"   db:"created_at"`
	UpdatedAt   time.Time    `json:"updatedAt"   db:"updated_at"`
}

type Ingredient struct {
	Name       string    `json:"name"`
	Quantity   *Quantity `json:"quantity"` // null for "to taste" ingredients
	IsOptional bool      `json:"isOptional"`
	Notes      string    `json:"notes"`
}

type Quantity struct {
	Min  *float64 `json:"min"`  // nullable for cases like "1 cup" (only max specified)
	Max  *float64 `json:"max"`  // nullable for cases like "at least 2 tbsp" (only min specified)
	Unit string   `json:"unit"` // required - tablespoons, grams, cups, cloves, etc.
}

type RecipeJob struct {
	ID           uuid.UUID  `json:"id"           db:"id"`
	UserID       uuid.UUID  `json:"userId"       db:"user_id"`
	URL          string     `json:"url"          db:"url"`
	Status       string     `json:"status"       db:"status"`
	ErrorMessage *string    `json:"errorMessage" db:"error_message"`
	RecipeID     *uuid.UUID `json:"recipeId"     db:"recipe_id"`
	NoteID       *uuid.UUID `json:"noteId"       db:"note_id"`
	CreatedAt    time.Time  `json:"createdAt"    db:"created_at"`
	CompletedAt  *time.Time `json:"completedAt"  db:"completed_at"`
}

type RecipeURLCache struct {
	URLHash      string    `json:"urlHash"     db:"url_hash"`
	OriginalURL  string    `json:"originalUrl" db:"original_url"`
	RecipeID     uuid.UUID `json:"recipeId"    db:"recipe_id"`
	CreatedAt    time.Time `json:"createdAt"   db:"created_at"`
	LastAccessed time.Time `json:"lastAccessed" db:"last_accessed"`
}

File: org-frontend/src/api/model/recipe.ts
// eslint-disable-next-line no-restricted-imports
import dayjs, { Dayjs } from "dayjs"

export interface Recipe {
  id: string
  name: string
  summary?: string
  servings?: number
  prepTime?: string
  sourceUrl?: string
  ingredients: Ingredient[]
  steps: string[]
  createdAt: Dayjs
  updatedAt: Dayjs
}

export interface Ingredient {
  name: string
  quantity?: Quantity
  isOptional: boolean
  notes: string
}

export interface Quantity {
  min?: number
  max?: number
  unit: string
}

export interface RecipeJob {
  id: string
  userId: string
  url: string
  status: "pending" | "processing" | "completed" | "failed"
  errorMessage?: string
  recipeId?: string
  noteId?: string
  createdAt: Dayjs
  completedAt?: Dayjs
}

export interface CreateRecipeRequest {
  url: string
}

export interface CreateRecipeResponse {
  jobId: string
}

export interface RecipeJobResponse {
  job: RecipeJob
  recipe?: Recipe
  note?: {
    id: string
    title: string
    content: string
  }
}

export function fromRecipeJson(recipe: Recipe): Recipe {
  return {
    ...recipe,
    createdAt: dayjs(recipe.createdAt),
    updatedAt: dayjs(recipe.updatedAt),
  }
}

export function fromJobJson(job: RecipeJob): RecipeJob {
  return {
    ...job,
    createdAt: dayjs(job.createdAt),
    completedAt: job.completedAt ? dayjs(job.completedAt) : undefined,
  }
}

export function fromJobResponseJson(
  response: RecipeJobResponse
): RecipeJobResponse {
  return {
    ...response,
    job: fromJobJson(response.job),
    recipe: response.recipe ? fromRecipeJson(response.recipe) : undefined,
  }
}
